### ISSUE TITLE
Duplicate constant declarations crash build process

### ISSUE THREAD

**Issue by jamesmillerburgess** [meteor/meteor#10052]
[Reproduction](https://github.com/jamesmillerburgess/meteor-duplicate-constants)

Although meteor tolerates a lot of malformed javascript, duplicate constant declarations result in the process crashing and having to restart from CLI.

It would be great if the process could wait on edits to try again instead of crashing. This happens a lot when I import something to replace a stub previously declared as a constant. Then if I don't have the terminal open to see the error, I stare at my browser waiting for it to refresh! 😐😐😐

```
/Users/james/Documents/GitHub/meteor/dev_bundle/lib/node_modules/meteor-promise/promise_server.js:218
      throw error;
      ^

TypeError: /Users/james/Documents/GitHub/test/client/main.js: Duplicate declaration "a"
  1 | const a = 1;
> 2 | const a = 2; // <-- Duplicate constants crash the build process
    |       ^
  3 |
  4 | // {{ <-- Malformed javascript like this results in 'Your application has errors. Waiting for file change'
  5 |
    at File.buildCodeFrameError (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transformation/file/file.js:231:12)
    at Scope.checkBlockScopedCollisions (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/scope/index.js:347:27)
    at Scope.registerBinding (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/scope/index.js:506:16)
    at Scope.registerDeclaration (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/scope/index.js:446:14)
    at Object.BlockScoped (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/scope/index.js:189:28)
    at Object.newFn (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/visitors.js:230:17)
    at NodePath._call (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:36:14)
    at NodePath.visit (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:88:12)
    at TraversalContext.visitQueue (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:118:16)
    at TraversalContext.visitMultiple (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:85:17)
    at TraversalContext.visit (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:144:19)
    at Function.traverse.node (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/index.js:94:17)
    at traverse (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/index.js:76:12)
    at NodePath.traverse (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/index.js:168:24)
    at Scope.crawl (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/scope/index.js:686:10)
    at Scope.init (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/scope/index.js:636:32)
    at NodePath.setScope (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:126:30)
    at NodePath.setContext (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:141:8)
    at new File (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transformation/file/file.js:86:8)
    at normalizeFile (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transformation/normalize-file.js:112:10)
    at runSync (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transformation/index.js:44:43)
    at Object.transformSync (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transform.js:43:38)
    at transform (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:103:26)
    at compile (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:122:5)
    at Cache.Cp.get (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/cache.js:101:26)
    at Object.exports.compile (/Users/james/Documents/GitHub/test/.meteor/local/isopacks/ecmascript/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:51:7)
    at Object.compile (packages/babel-compiler.js:44:29)
    at packages/babel-compiler.js:210:22
    at Function.time (/tools/tool-env/profile.js:305:10)
    at profile (packages/babel-compiler.js:258:20)
    at BabelCompiler.BCp.processOneFileForTarget (packages/babel-compiler.js:209:20)
    at packages/babel-compiler.js:123:25
    at /Users/james/Documents/GitHub/meteor/dev_bundle/lib/node_modules/meteor-promise/fiber_pool.js:43:40
 => awaited here:
    at Promise.await (/Users/james/Documents/GitHub/meteor/dev_bundle/lib/node_modules/meteor-promise/promise_server.js:60:12)
    at JsOutputResource.finalize (/tools/isobuild/compiler-plugin.js:887:7)
    at JsOutputResource.hasPendingErrors (/tools/isobuild/compiler-plugin.js:914:10)
    at JsOutputResource.reportPendingErrors (/tools/isobuild/compiler-plugin.js:919:14)
    at ImportScanner._scanFile (/tools/isobuild/import-scanner.js:823:14)
    at outputFiles.forEach.file (/tools/isobuild/import-scanner.js:503:14)
    at Array.forEach (<anonymous>)
    at ImportScanner.scanImports (/tools/isobuild/import-scanner.js:501:22)
    at sourceBatches.forEach.batch (/tools/isobuild/compiler-plugin.js:1277:17)
    at Array.forEach (<anonymous>)
    at Function.computeJsOutputFilesMap (/tools/isobuild/compiler-plugin.js:1245:19)
    at ClientTarget._emitResources (/tools/isobuild/bundler.js:1111:8)
    at buildmessage.enterJob (/tools/isobuild/bundler.js:840:12)
    at /tools/utils/buildmessage.js:359:18
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:352:34
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:350:23
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.enterJob (/tools/utils/buildmessage.js:324:26)
    at ClientTarget.make (/tools/isobuild/bundler.js:828:18)
    at /tools/isobuild/bundler.js:3066:14
    at /tools/isobuild/bundler.js:3155:20
    at Array.forEach (<anonymous>)
    at Function._.each._.forEach (/Users/james/Documents/GitHub/meteor/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
    at /tools/isobuild/bundler.js:3154:7
    at /tools/utils/buildmessage.js:271:13
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:3047:31)
    at files.withCache (/tools/isobuild/bundler.js:2994:32)
    at Object.withCache (/tools/fs/files.js:1710:12)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2994:16)
    at Profile.run (/tools/runners/run-app.js:579:36)
    at Function.run (/tools/tool-env/profile.js:490:12)
    at bundleApp (/tools/runners/run-app.js:578:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:622:35)
    at AppRunner._fiber (/tools/runners/run-app.js:880:28)
    at /tools/runners/run-app.js:408:12
```

**Comment by fabien-h** [comment]
I would prefer to keep meteor crashing by default when you have obvious mistakes like that one.

In this case, babel is the one crashing meteor. Maybe you can try to edit your babelrc file. (but a const is not supposed to be redefined or duplicated so I'm not even sure that the option exists)

**Comment by benjamn** [comment]
@fabien-h I think the request is that Meteor should display the error and wait for you to fix it, rather than shutting down completely, thus forcing you to restart the build tool from scratch. Obvious mistakes would still be obvious; you could just fix them faster.

**Comment by aliogaili** [comment]
I'm seeing the same issue as well, it's crashing the whole server forcing a restart which interrupt the dev process..

**Comment by pisti2010** [comment]
+1 to an error state vs a full crash. It slows down dev process to have to rebuild for quickly solved mistakes (duplicate imports is a common one)

**Comment by herchu** [comment]
+1 ! 
The issue forces to restart the server and tooling, so much slower than when any other type of "compiler" error is found. It should be treated, if possible, as any other syntax error (stop and wait for file changes)

**Comment by benjamn** [comment]
This should be fixed in Meteor 1.8, with `babel-compiler@7.2.0`, thanks to this commit: https://github.com/meteor/meteor/commit/bbdf28544f3bae08bfdd01e58ebb37ff04c4b521. Please feel free to reopen this issue if the problem persists after updating to Meteor 1.8.
